<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование таблиц</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица <code>books</code></a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица <code>authors</code></a></span></li><li><span><a href="#Таблица--publishers" data-toc-modified-id="Таблица--publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица  <code>publishers</code></a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица <code>ratings</code></a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица <code>reviews</code></a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li></ul></div>

# Анализ базы данных средствами SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Ваша компания решила быть на волне и купила крупный сервис для чтения книг
по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица** `books`

Содержит данные о книгах:
* `book_id` - идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

**Таблица** `authors`

Содержит данные об авторах:
* `author_id` — идентификатор автора;
* `author` — имя автора.

**Таблица** `publishers`

Содержит данные об издательствах:
* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства.

**Таблица** `ratings`

Содержит данные о пользовательских оценках книг:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

**Таблица** `reviews`

Содержит данные о пользовательских обзорах на книги:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, написавшего обзор;
* `text` — текст обзора

## Исследование таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
# создадим функцию, которая будет возвращать первые 10 строк, а также общие сведения нужной нам таблицы
def information(data):
    display(data.head(10))  
    data.info()

### Таблица `books`


In [3]:
query = '''
SELECT * FROM BOOKS 
'''
books = pd.io.sql.read_sql(query, con = engine) 
information(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


### Таблица `authors`

In [4]:
query = '''
SELECT * FROM AUTHORS 
'''
authors = pd.io.sql.read_sql(query, con = engine) 
information(authors)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### Таблица  `publishers`

In [5]:
query = '''
SELECT * FROM PUBLISHERS 
'''
publishers = pd.io.sql.read_sql(query, con = engine) 
information(publishers)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


### Таблица `ratings`

In [6]:
query = '''
SELECT * FROM RATINGS 
'''
ratings = pd.io.sql.read_sql(query, con = engine) 
information(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


### Таблица `reviews`

In [7]:
query = '''
SELECT * FROM REVIEWS 
'''
reviews = pd.io.sql.read_sql(query, con = engine) 
information(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## Ответы на вопросы

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [8]:
query = '''
SELECT COUNT(BOOK_ID) FROM BOOKS
WHERE PUBLICATION_DATE >= '2000-01-01' 
'''
pd.io.sql.read_sql(query, con = engine) 

count
0    821

**Ответ:** после 1 января 2000 года вышла 821 книга.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [9]:
query = '''
SELECT B.BOOK_ID, B.TITLE, REV.COUNT_OF_REVIEW, RAT.AVG_RATING FROM BOOKS B 
LEFT JOIN (SELECT R.BOOK_ID, COUNT(R.REVIEW_ID) COUNT_OF_REVIEW FROM REVIEWS R
GROUP BY R.BOOK_ID) REV ON B.BOOK_ID = REV.BOOK_ID
LEFT JOIN (SELECT R.BOOK_ID, AVG(R.RATING) AVG_RATING FROM RATINGS R
GROUP BY R.BOOK_ID) RAT ON B.BOOK_ID = RAT.BOOK_ID
ORDER BY RAT.AVG_RATING
'''
pd.io.sql.read_sql(query, con = engine) 

book_id                                              title  \
0        303                               Harvesting the Heart   
1        202                                      Drowning Ruth   
2        371                                              Junky   
3        316                  His Excellency: George Washington   
4        915  The World Is Flat: A Brief History of the Twen...   
..       ...                                                ...   
995      182                                         Dead Souls   
996      599  Stone of Farewell (Memory  Sorrow  and Thorn  #2)   
997       62                                      Alas  Babylon   
998      913                                 The Woman in Black   
999      441                  Misty of Chincoteague (Misty  #1)   

     count_of_review  avg_rating  
0                2.0        1.50  
1                3.0        2.00  
2                2.0        2.00  
3                2.0        2.00  
4                3.0        2.25  
..               ...         ...  
995              2.0        5.00  
996              1.0        5.00  
997              2.0        5.00  
998              2.0        5.00  
999              2.0        5.00  

[1000 rows x 4 columns]

**Ответ:** глядя на результат, можно сказать, что для многих книг было написано больше одного обзора. Минимальная средняя ценка книги - 1.5, максимальная - 5.

###  Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [10]:
query = '''
SELECT B.PUBLISHER_ID, P.PUBLISHER, COUNT(B.BOOK_ID) COUNT_OF_BOOKS FROM BOOKS B
INNER JOIN PUBLISHERS P ON 
B.PUBLISHER_ID = P.PUBLISHER_ID
WHERE B.NUM_PAGES > 50
GROUP BY B.PUBLISHER_ID, P.PUBLISHER
ORDER BY COUNT(B.BOOK_ID) DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine) 

publisher_id      publisher  count_of_books
0           212  Penguin Books              42

**Ответ:** больше всего книг выпустило издательство ***Penguin Books*** - на их счету 42 книги.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
query = '''
SELECT A.AUTHOR, ROUND(AVG(RAT.AVG_RATING), 2) AVG_RAT FROM AUTHORS A 
INNER JOIN BOOKS B
ON A.AUTHOR_ID = B.AUTHOR_ID
INNER JOIN 
(SELECT R.BOOK_ID, AVG(R.RATING) AS AVG_RATING FROM RATINGS R
GROUP BY R.BOOK_ID
HAVING COUNT(R.RATING_ID) >= 50) RAT
ON B.BOOK_ID = RAT.BOOK_ID
GROUP BY A.AUTHOR
ORDER BY AVG_RAT DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine) 


author  avg_rat
0  J.K. Rowling/Mary GrandPré     4.28

**Ответ:** автор с самой высокой средней оценкой книг - ***Джоан Роулинг*** в соавторстве с ***Мэри Гранпрэ***.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''
SELECT AVG(COUNT_50_REVIEWS.COUNT_REVIEWS) FROM 
(SELECT COUNT(R.REVIEW_ID) COUNT_REVIEWS, R.USERNAME FROM REVIEWS R
INNER JOIN 
(SELECT USERNAME FROM RATINGS
GROUP BY USERNAME
HAVING COUNT(RATING) > 50) RAT
ON R.USERNAME = RAT.USERNAME
GROUP BY R.USERNAME) COUNT_50_REVIEWS
'''
pd.io.sql.read_sql(query, con = engine) 


avg
0  24.333333

**Ответ:** в среднем пользователи, поставившие более 50 оценок, оставляют ***24 отзыва***.